### PR TITLE
allow github reporter to work with git@ urls

### PR DESCRIPTION
**Pull request by dragon512**
Allow GitHub to parse https:// and git@github.com:xxx style repository URLS correctly
Add some verbose messages that can be useful.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/reporters/github.py [hunk 1]
```diff
@@ -135,12 +142,24 @@ def send(self, build):
         else:
             issue = None
 
-        if project:
+        if "/" in project:
             repoOwner, repoName = project.split('/')
-        else:
+        # Given we have a repository define try to get the owner and name
+        elif sourcestamps[0]['repository']:
```
Comment (tardyp) [inline master/buildbot/reporters/github.py]: Please install codecov chrome extension or look at the report.
We require that those kind of changes are unit tested before they can be included in the code

File: master/buildbot/reporters/github.py [hunk 2]
```diff
@@ -94,10 +95,16 @@ def createStatus(self,
         if context is not None:
             payload['context'] = context
 
-        return self._http.post(
+        if self.verbose:
```
Comment (tardyp) [inline master/buildbot/reporters/github.py]: I guess this is a separate feature which belongs to a separate PR. I think this would be better to put it in HttpBase reporter

File: master/buildbot/reporters/github.py [hunk 3]
```diff
@@ -135,12 +142,24 @@ def send(self, build):
...
+        if "/" in project:
             repoOwner, repoName = project.split('/')
-        else:
+        # Given we have a repository define try to get the owner and name
+        elif sourcestamps[0]['repository']:
             repo = sourcestamps[0]['repository'].split('/')[-2:]
-            repoOwner = repo[0]
-            repoName = '.'.join(repo[1].split('.')[:-1])
+            # if this repo is in the form of git@ not http://
+            if sourcestamps[0]['repository'].startswith("git@"):
```
Comment (tardyp) [inline master/buildbot/reporters/github.py]: I think this algorithm is useful not only for github, but also for bitbucket, gitlab, and all github clones.

We should factorize it in an utility function in (create new buildbot.util.repository_urls module)
```
def getUserAndProjectFromRepoUrl(repoUrl):
     [...]
```

File: master/buildbot/reporters/github.py (hunk 1 above)
Comment (noc0lour) [inline master/buildbot/reporters/github.py]: Maybe using the `project` property is safe since github changes coming in via the www hook or `PullrequestPoller` have the project property set to the correct value only timed/force schedulers would have to set that property.

**Comment by Frodox** [comment]
Looks like it should fix #3215 
Ping @dragon512 @samvasko

**Comment by dragon512** [comment]
Cool, I should pull this change in and redo the fix with this new util.

**Comment by tardyp** [comment]
@dragon512 please do.

File: master/buildbot/reporters/github.py [hunk 4]
```diff
@@ -205,5 +216,6 @@ def createStatus(self,
         payload = {'body': description}
 
         return self._http.post(
-            '/'.join(['/repos', repo_user, repo_name, 'issues', issue, 'comments']),
+            '/'.join(['/repos', repo_user, repo_name,
```
Comment (Frodox) [inline master/buildbot/reporters/github.py]: What is the goal of such wrapping here, and leave [this one](https://github.com/dragon512/buildbot/blob/bb83f28def75730132d4d6709c0630009e70dd1d/master/buildbot/reporters/github.py#L188) as is? (Since it is less 80 characters anyway)

Comment (dragon512) [inline master/buildbot/reporters/github.py]: I formatted with autopep8. It seems to want to do this. I don't know why. I switch it back and did the format again .. this went back to the two line statement.

I can see about making the line length longer and seeing it this will go away

Comment (dragon512) [inline master/buildbot/reporters/github.py]: Ya I just tried again with --max-line-length 132
It was still reformatting this line. I can manually set it back

**Comment by dragon512** [comment]
@tardyp I am unclear how this is failing worse. I rebased. I have a new test added to cover the basic three lines that had not been covered before. How is this reporting even worse coverage than before?

File: master/buildbot/reporters/github.py [hunk 5]
```diff
@@ -136,12 +137,16 @@ def send(self, build):
...
         else:
-            repo = sourcestamps[0]['repository'].split('/')[-2:]
-            repoOwner = repo[0]
-            repoName = '.'.join(repo[1].split('.')[:-1])
+            giturl = giturlparse(sourcestamps[0]['repository'])
+            repoOwner = giturl.owner
+            repoName = giturl.repo
+
+        if self.verbose:
+            log.msg("Updating github status: repoOwner={repoOwner}, repoName={repoName}".format(
```
Comment (tardyp) [inline master/buildbot/reporters/github.py]: codecov says this line is not covered. that's fine to me. just fix the doc issue.

**Comment by dragon512** [comment]
there is no real doc change, as this fixes the reporter to correct consume json data with git@ formatted URLs. this is more of a bug fix, vs a API change

**Comment by Frodox** [comment]
@dragon512 @tardyp means doc *issus*, i.e. it doesn build because of spelling. [Take a look test](https://nine.buildbot.net/#/builders/7/builds/891)

> relnotes/index.rst:33:RepoName:["Forename", "Repairmen", "Namedrop", "Reamonn", "Repentance", "Repiner"]
relnotes/index.rst:62:RepoName:["Forename", "Repairmen", "Namedrop", "Reamonn", "Repentance", "Repiner"]

File: master/buildbot/newsfragments/github.feature [hunk 6]
```diff
@@ -0,0 +1 @@
+GitHubStatusPush now support reporting to ssh style URLs, ie git@github.com:Owner/RepoName.git
```
Comment (Frodox) [inline master/buildbot/newsfragments/github.feature]: Could you please add link on GitHubStatusPush, format it with backtricks, and backtricks for `git@github...` (mark it as code). And looks like spell check doesn't like this `RepoName`

File: master/buildbot/reporters/github.py [hunk 7]
```diff
@@ -136,12 +137,16 @@ def send(self, build):
         else:
             issue = None
 
-        if project:
+        if "/" in project:
```
Comment (Frodox) [inline master/buildbot/reporters/github.py]: what is it for? who can set `project` to smth like `repoOwner/repoName` ?

Comment (dragon512) [inline master/buildbot/reporters/github.py]: This was for when I set project to "foobar"  not "owner/repo" This protects the code and allow the code to use the meta data that was really sent to get the data. There is no documentation other than that the project is a meta value. there is nothing saying that it can have magic values. I been burned by this a few times.

File: master/buildbot/newsfragments/github.feature (hunk 6 above)
Comment (dragon512) [inline master/buildbot/newsfragments/github.feature]: Thanks! I think I corrected it properly

**Comment by dragon512** [comment]
@tardyp I think I have it. Sorry for the delay on finishing this.

**Comment by dragon512** [comment]
@tardyp Just a friendly ping. Is there anything holding back this PR from being merged?
